Implement InstanceExistsByProviderID() for cloud providers

FengyunPan opened the pull request:
Fix #51406
If cloud providers(like aws, gce etc...) implement ExternalID()
and support getting instance by ProviderID , they also implement
InstanceExistsByProviderID().

/assign wlan0
/assign @luxas

**Release note**:
```release-note
NONE
```

FengyunPan commented:
/test pull-kubernetes-federation-e2e-gce

luxas commented:
@wlan0 could you please take a look at this PR? I'm gonna be out for a few days

wlan0 commented:
@FengyunPan Have you tested all the cloudproviders? The code change looks good.

FengyunPan commented:
@wlan0 No, not all, I have tested openstack cloud provider, I have not other cloud to test.
I am sorry for that, my codes is very simple and is easy to read, so I submit it.

wlan0 commented:
I'm ok with the code changes. @luxas are you ok with merging without testing all the cloudproviders? 

@cheftako @jdumars @luomiao @justinsb can you review the implementation of InstanceExistsByProviderID for your clouds and let us know if the change LGTY? 

Thanks!

jdumars commented:
@brendandburns @lachie83 @seanknox PTAL

cheftako commented on pkg/cloudprovider/providers/gce/gce_instances.go:
```diff
@@ -157,7 +156,20 @@ func (gce *GCECloud) ExternalID(nodeName types.NodeName) (string, error) {
+		return false, err
+	}
+
+	_, err = gce.getInstanceFromProjectInZoneByName(project, zone, name)
```
Can we create an instanceByProviderID method? Then both InstanceExistsByProviderID and InstanceTypeByProviderID can both call that.

FengyunPan replied:
Ok, done.

luomiao commented:
The change at vsphere cloud provider looks good to me.
@BaluDontu @divyenpatel can you also please verify since these functions are recently updated.

BaluDontu commented:
@FengyunPan : I was just trying to figure out what is the need for InstanceExistsByProviderID() when kubernetes can call InstanceID() to know the status of a node?

FengyunPan commented:
@BaluDontu That's no need. I update it soon.

FengyunPan commented:
/test pull-kubernetes-e2e-gce-bazel

wlan0 commented:
/test pull-kubernetes-e2e-gce-bazel

divyenpatel commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -379,7 +379,25 @@ func (vs *VSphere) ExternalID(nodeName k8stypes.NodeName) (string, error) {
+		return false, err
+	}
+
+	return true, nil
```
Do we want to consider the power state of the Node VM here? Node VM can be obtained from the Inventory but it can be in the powered off state. Powered off Node VM should be immediately deleted by the cloud controller manager.

wlan0 replied:
Great catch! If it should be deleted, then that "Power" state should be checked for, and "false" should be returned if the NodeVM is "powered off". @FengyunPan

BaluDontu replied:
@jingxu97 : Can you please comment on this. Should the node be deleted by controller manager if a node is powered off?

FengyunPan replied:
@divyenpatel @wlan0 @BaluDontu I do not think we should delete "Power off" state node immediately. In this function, we just check whether node is exist.
Please look this comment: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L230
@divyenpatel  I see your issuse, we should  consider  "Power off" state node, but not here, we can not delete "Power off" state node immediately, it is unsafe.
And we have a discuss ( #46442), the discuss is still underway, please take a look and comments at there.

BaluDontu commented:
@FengyunPan : From vsphere side, the change looks good to us. Its an lgtm from our side.

FengyunPan commented:
@cheftako @jdumars @justinsb @brendandburns @lachie83 @seanknox PTAL
Any thoughts on this PR?

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1104,7 +1104,24 @@ func (c *Cloud) ExternalID(nodeName types.NodeName) (string, error) {
+		return false, nil
+	}
+
+	return true, nil
```
I think you need to check the running state.

Also, this duplicates logic in `ExternalID`, so I propose one calls into the other.

justinsb replied:
On second thoughts sharing code looks complicated, but the logic shouldn't change IMO

justinsb commented:
AWS needs to check the running state I believe

FengyunPan commented:
In this function, we just check whether node is exist.
Please look this comment: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L230
But I can't understand why the comment of 'InstanceExistsByProviderID' is need to check the running state. They are conflict, and is hard to review for reviewers.
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L1147

By the way, if 'InstanceExistsByProviderID' should check the running state, why not named ''InstanceActiveByProviderID'. If instance is exist and not running state, should 'false, error' be returned?

IMO, we just check whether node is exist, let (#46442) to check running state and do some right things

@justinsb @luxas @wlan0 Any thoughts on this?

justinsb commented:
We definitely have to filter instances in the terminated state.

I think we should address the stopped state in 1.9. I don't think we should have different behaviors for the external vs the included cloud controller manager. But that is just my opinion, and happy to discuss.. but it should be a deliberate and documented decision.

wlan0 commented:
@FengyunPan Those are good points you've raised. In addition to what you've mentioned, the behavior of this implementation should be exactly the same as the existing included cloud providers.

Moreover, turned off nodes cannot be used to schedule pods or mount volumes, but it takes up KCM/CCM resources anyways.

> We definitely have to filter instances in the terminated state.

I agree with @justinsb that turned off nodes should be filtered for now.

FengyunPan commented:
@justinsb @wlan0 
Oh, I see. Ok, I will update this PR after the discuss of (#46442).
Thank for your comment very much.

cheftako reviewed (approved):
/lgtm

jdumars commented:
I'm trying to get the Azure perspective reviewed on this.  Please do not merge until that review is completed.

itowlson commented on pkg/cloudprovider/providers/azure/azure_instances.go:
```diff
@@ -87,10 +86,23 @@ func (az *Cloud) ExternalID(name types.NodeName) (string, error) {
 }
 
-// InstanceExistsByProviderID returns true if the instance with the given provider id still exists and is running.
+// InstanceExistsByProviderID returns true if the instance with the given provider id still exists.
```
The `az.InstanceID` function says it returns `InstanceNotFound` if the instance is not running.  So perhaps the "and is running" part of the documentation needs to be reinstated?

FengyunPan replied:
Right, thank you.

itowlson commented on pkg/cloudprovider/providers/azure/azure_instances.go:
```diff
@@ -87,10 +86,23 @@ func (az *Cloud) ExternalID(name types.NodeName) (string, error) {
+		return false, err
+	}
+
+	return true, nil
```
Is it possible to add tests for this function?

FengyunPan replied:
It is too simple and there is no good test case for this function.

FengyunPan commented:
@justinsb I have checked the status of AWS instance. PTAL, thank you.
@BaluDontu @divyenpatel I have checked the status of VSphere instance, PTAL, thank you.

luxas commented:
@jdumars @brendandburns Please review from the Azure side. Thanks!

FengyunPan commented:
/retest

FengyunPan commented:
ping @jdumars 
ping @brendandburns 
ping @itowlson 
PTAL, thank you.

FengyunPan2 committed:
Implement InstanceExistsByProviderID() for cloud providers

Fix #51406
If cloud providers(like aws, gce etc...) implement ExternalID()
and support getting instance by ProviderID , they also implement
InstanceExistsByProviderID().

itowlson commented:
Azure implementation looks okay to me.

FengyunPan commented:
ping @luxas @wlan0 
Is it ok to approve?

wlan0 commented:
@FengyunPan yes

/lgtm
/approve

FengyunPan commented:
/test pull-kubernetes-unit

discordianfish commented:
It looks like this isn't working in my case and instances get deleted. The providerID looks correct and everything, yet every new node I add gets deleted right away without any error. The only thing the controller-manager logs is:
```
E1204 10:55:40.596096       1 actual_state_of_world.go:483] Failed to set statusUpdateNeeded to needed true because nodeName="ip-172-20-58-74.ec2.internal"  does not exist
E1204 10:55:40.596332       1 actual_state_of_world.go:497] Failed to update statusUpdateNeeded field in actual state of world: Failed to set statusUpdateNeeded to needed true because nodeName="ip-172-20-58-74.ec2.internal"  does not exist
```

While in the events log I see (different node but same result):
```
24s         24s          1         ip-172-20-79-61.ec2.internal.14fd183319cda415     Node                  Normal    DeletingNode              controllermanager                         Node ip-172-20-79-61.ec2.internal event: Deleting Node ip-172-20-79-61.ec2.internal because it's not present according to cloud provider
```

While this is the node object before it got removed: https://gist.github.com/discordianfish/f6ba957db7cb8e2875a8bea0505a5097

I'm going to downgrade kubernetes to see if this fixes it.

dims commented:
@discordianfish sounds serious for 1.9, can you please open a bug so we can track it better?

discordianfish commented:
@dims Turned out I just missed the KubernetesCluster tags but it took quite long to figure that out due to lack of documentation of the cloud provider itself. The UX is pretty bad, with the cloud intergration just deleting your nodes without giving any clue what might be going on. Would be great if the error returned the filters it used to search for matching instances.

dims commented:
@discordianfish thanks for tracking it down. yes good news and bad news :)

cc @justinsb
